reactor-pool/src/main/java/reactor/pool/SimpleDequePool.java

```diff
@@ -372,89 +375,97 @@ private void drainLoop() {
+							Mono<POOLABLE> allocator = allocatorWithScheduler();
+
+							Mono<POOLABLE> primary = allocator.doOnEach(sig -> {
+								if (sig.isOnNext()) {
+									POOLABLE newInstance = sig.get();
+									assert newInstance != null;
+									ACQUIRED.incrementAndGet(this);
+									metricsRecorder.recordAllocationSuccessAndLatency(clock.millis() - start);
+									borrower.deliver(createSlot(newInstance));
+								}
+								else if (sig.isOnError()) {
+									Throwable error = sig.getThrowable();
+									assert error != null;
+									metricsRecorder.recordAllocationFailureAndLatency(clock.millis() - start);
+									poolConfig.allocationStrategy()
+											.returnPermits(1);
+									borrower.fail(error);
+								}
+							}).contextWrite(borrower.currentContext());
 
-						if (permits == 1) {
-							//subscribe to the primary, which will directly feed to the borrower
-							primary.subscribe(alreadyPropagated -> { }, alreadyPropagatedOrLogged -> drain(), this::drain);
-						}
-						else {
-							/*=============================================*
-							 * (warm up in sequence to primary allocation) *
-							 *=============================================*/
-							int toWarmup = permits - 1;
-							logger.debug("should warm up {} extra resources", toWarmup);
-
-							final long startWarmupIteration = clock.millis();
-							// flatMap will eagerly subscribe to the allocator from the current thread, but the concurrency
-							// can be controlled from configuration
-							final int mergeConcurrency = Math.min(poolConfig.allocationStrategy().warmupParallelism(), toWarmup + 1);
-							Flux.range(1, toWarmup)
-									.map(i -> warmupMono(i, toWarmup, startWarmupIteration, allocator).doOnSuccess(__ -> drain()))
-									.startWith(primary.doOnSuccess(__ -> drain()).then())
-									.flatMap(Function.identity(), mergeConcurrency, 1) // since we dont store anything the inner buffer can be simplified
-									.onErrorResume(e -> Mono.empty())
-									.subscribe(aVoid -> { }, alreadyPropagatedOrLogged -> drain(), this::drain);
+							if (permits == 1) {
+								//subscribe to the primary, which will directly feed to the borrower
+								primary.subscribe(alreadyPropagated -> { }, alreadyPropagatedOrLogged -> drain(), this::drain);
+							}
+							else {
+								/*=============================================*
+								 * (warm up in sequence to primary allocation) *
+								 *=============================================*/
+								int toWarmup = permits - 1;
+								logger.debug("should warm up {} extra resources", toWarmup);
+
+								final long startWarmupIteration = clock.millis();
+								// flatMap will eagerly subscribe to the allocator from the current thread, but the concurrency
+								// can be controlled from configuration
+								final int mergeConcurrency = Math.min(poolConfig.allocationStrategy().warmupParallelism(), toWarmup + 1);
+								warmupInProgress = true;
```

Comment:
This flag does not guarantee exclusivity IMO. A stress test could be added to validate my accusation and I worry it would hold true :(

Reply:
I also think so

Reply:
I do not see for the moment the concurrency issue, but I will give investigate this, and will create a stress test.
The flag is only set from the drainLoop and then it is reset once the subscription to the Flux.range completes, and if drainLoop is missing the update of warmupInProgress, then drain() will cause drainLoop to be called again.
However, since both of you are seeing an issue here, I'll carefully revisit all this.

There is however something that I missed: it is when the subscription to the `primary` is failing, in this case there is a problem because the flag is not reset to false.

Another PR is addressing this problem: https://github.com/reactor/reactor-pool/pull/175